### PR TITLE
[WIP] bootstrapped TypeScript declarations

### PR DESCRIPTION
**Pull request by hzlmn**
cc @lukeed @jbucaran 

Hi, as discussed.

Also i am wondering how we could describe node related functions like `read` `write` that works with `Buffer` object ? Should we pass `@typings/node` as `peerDependency` ?

### CODE REVIEW

**Comment by jorgebucaran** [comment]
@hzlmn Woah, this is awesome thank you! 🤘 

> Should we pass @typings/node as peerDependency ?

What would that be used for? I thought typings are mostly there to help code editors offer better completion, help type linters to check for type mismatches, etc.

**Comment by lukeed** [comment]
Nice @hzlmn! I'm not sure what the answers are to your questions. I don't use TS but can help you find the answer if you need help.

Also, do you mind restructuring a little bit? Can we move this to the project root as `fly.d.ts`?

Then you also have to add `fly.d.ts` to the `files` entry of the `package.json`, or else your hard work won't be downloaded.

**Comment by hzlmn** [comment]
@jbucaran Yes, they are. But as fly is node tool it use some Node APIs like `Buffer`, for example, that is unknown for TypeScript by default,  but there are separate declarations for `Node` [`@typings/node`](https://www.npmjs.com/package/@types/node)

@lukeed Yes sure, restructuring is good, that's why i create this as a `WIP`

Possible solutions is to add `@types/node` as a peerDependency [like here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express-serve-static-core/index.d.ts) or just define `any` type for Node related stuffs but it is not very convenient.

**Comment by lukeed** [comment]
Thanks!

Hmm.. It's probably not ideal, but I think the `any` type might be better here, because we don't want to tell every downloader that they're missing a `peerDep` for `@types/node`. I, for one, would be annoyed by that since I don't use TS.

File: package.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "lib",
     "cli.js"
   ],
+  "typings": "./d.ts/index.d.ts",
```
Comment (jorgebucaran) [inline package.json]: @hzlmn What's this part for? Just trying to understand how this works.

Comment (hzlmn) [reply package.json]: @jbucaran Sure, see [here](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html). It is just an alias for `types` that allows TypeScript's compiler to get suitable declaration file for package. If you won't do it, you need manually load your declaration from package with triple-slash directive.

Comment (jorgebucaran) [reply package.json]: Thanks @hzlmn 👍  

What if I am not using TypeScript and just want my editor to pickup the types for intelligent code suggestions?

Comment (hzlmn) [reply package.json]: @jbucaran I don't know about other editors but i am pretty sure that vscode and webstorm can do that.

Comment (jorgebucaran) [reply package.json]: @hzlmn Interesting!

Comment (lukeed) [reply package.json]: @hzlmn @jbucaran I believe all editors' typescript plugins _do_ pickup the `types` field. This is what allows all of them to read definitions & provide completions.

Comment (jorgebucaran) [reply package.json]: @lukeed 🤔 Did you mean `typings` field?

Comment (lukeed) [reply package.json]: Oops, yep 😝

**Comment by hzlmn** [comment]
Sorry, forgot about it. Updated

**Pull request closed without merging**